### ISSUE TITLE
ifndef::foo,bar[] is not backwards compatible 

### ISSUE THREAD

**Issue by jjaderberg** [asciidoctor/asciidoctor#1983]
Using `ifdef` with multiple attributes is backwards compatible in Asciidoctor, but `ifndef` is not. The negation has different scope.

With Asciidoc and comma-separated attributes ("any"), the negation in `ifndef` is applied to the union: if any of the attributes is defined, then false. This means that if `ifdef` would have evaluated to true for the same comma-separated list of attributes, then `ifndef` is false.

In Asciidoctor, the negation in `ifndef` is distributed: if any of the attributes is not defined, then true. This means that `ifndef` is not negating `ifdef`--indeed they can both be true at the same time for the same list of attributes.

## Example AsciiDoc
 
````
= Ifndef if ifdef::asciidoctor[] diff from if ifndef::asciidoctor[]

ifdef::foo,bar[]
****
Foo.
****
endif::[]
ifndef::foo,bar[]
****
Bar.
****
endif::[]
````

Processing this document and defining one of the attributes gives different results with Asciidoc and Asciidoctor.

```
asciidoc -a foo foobar.adoc
````

will render the `Foo.` sidebar. `ifdef::foo,bar[]` is true because `foo` is defined. The `Bar.` sidebar is not rendered because `ifndef::foo,bar[]` is false, since the negation in `ifndef` is applied to the union--if any of the attributes is defined, then the expression as a whole is false.

Processing with AsciiDoctor

```
asciidoctor -a foo foobar.adoc
```

both `Foo.` and `Bar.` sidebars are rendered.

## What about `foo+bar`?

Using a plus-separated list of attributes instead ("all"), the behavior is reversed. Adding to the example document a `Foobar.` sidebar under a `ifndef::foo+bar[]` condition:

````
ifdef::foo,bar[]
****
Foo.
****
endif::[]
ifndef::foo,bar[]
****
Bar.
****
endif::[]
ifndef::foo+bar[]
****
Foobar.
****
endif::[]
````

Asciidoctor will still render the `Foo.` and `Bar.` sidebars, but not the `Foobar.` sidebar. Asciidoc renders the `Foo.` and `Foobar.` sidebars.

Is this an intentional breaking change? It seems unnecessary since what it amounts to is swapping the meaning of `ifndef::foo,bar[]` and `ifndef::foo+bar[]`.

**Comment by mojavelinux** [comment]
I vaguely remember changing the behavior of these conditionals. However, it was so long ago I really can't be sure whether or not this difference was intentional.

Here's an example that shows the logic of how it's implemented today:

```
ifdef::foo,bar[foo or bar is set]
ifdef::foo+bar[foo and bar are set]
ifndef::foo+bar[neither foo or bar are set (i.e., both foo and bar are unset)]
ifndef::foo,bar[one of foo or bar is not set]
```

Notice that the "ifndef" cases are not attempting to be a negated version of the "ifdef" cases. Instead, they are trying to assert whether the attributes are not set.

In other words, the "ifndef" with `+` asserts that both attributes are not set. The "ifndef" with `,` asserts that at least one is not set. You need to read it as "if not def" as opposed to "not if def".

If we decide to keep the current logic, we could preserve compliance with AsciiDoc Python by using the old logic if compat-mode is set. But before we do that, we need to decide whether we want to maintain the current logic.

**Comment by mojavelinux** [comment]
I'm proposing the following reading:

* `ifndef::foo+bar` - if foo and bar are not defined
* `ifndef::foo,bar` - if foo or bar are not defined

**Comment by mojavelinux** [comment]
I'm inclined to push this to 1.5.7 and possibly even 1.6.0. I'm not sure we're ready to make a decision yet.

The main concern I have is whether its possible to cover all scenarios with the current behavior. To my knowledge, they are covered. Therefore, this falls back to just a difference between AsciiDoc Python and Asciidoctor.

Do you think this can be addressed simply by documenting the behavior and the divergence from AsciiDoc Python?

**Comment by mojavelinux** [comment]
I've done *a lot* more thinking about this and I think in the end, you're right. It's very difficult to read it the way I was proposing.

The simplest way to check the logic is to replace "ifndef" with "unless" (ala Ruby). So if you think of "ifndef foo and bar" as "unless foo and bar", then you can recognize the logic to be "unless both foo and bar are defined".

**Comment by mojavelinux** [comment]
While this is a breaking change, I'm hard pressed to find any uses of logical operators combined with ifndef, so I think we're in the clear.
